### PR TITLE
spherical camera

### PR DESCRIPTION
**Pull request by lucacarlone**
Added spherical camera to model generic wide-field of view cameras.
Generalized triangulation functions to allow the development of arbitrary cameras.
Adjusted smart factors to remove pinhole assumption.

@dellaert : I suggest merging the PR "smart factors for arbitrary cameras" first. If possible, can you also add my student Marcus Abate mabate@mit.edu to GTSAM and add as reviewer here?

### CODE REVIEW

**Comment by lucacarlone** [comment]
@dellaert : this is done, but I suggest merging the PR "feature/rollingShutterSmartFactors" first. If possible, can you also add my student Marcus Abate mabate@mit.edu to GTSAM and add a reviewer here?

**Comment by varunagrawal** [comment]
Awesome. I was hoping to add spherical camera support in the future (did a lot of learning about rectification and the double sphere model this past summer).

@lucacarlone Would it be possible to reduce the noise by not having the extra unrelated commits?

**Comment by lucacarlone** [comment]
> Awesome. I was hoping to add spherical camera support in the future (did a lot of learning about rectification and the double sphere model this past summer).
> 
> @lucacarlone Would it be possible to reduce the noise by not having the extra unrelated commits?

thanks @varunagrawal ! I suggest reviewing PRs in this order:
- Feature/rolling shutter smart factors  feature #827
- Smart Factors for arbitrary cameras #862
- spherical camera  #861
which will highlight what each PR adds.
Let me know if that makes sense.

**Comment by dellaert** [comment]
I'll take a look at #827, which seems to be blocked on my templating comment. I'll check it out on my side and then merge if I can't fix it.

**Comment by lucacarlone** [comment]
> I'll take a look at #827, which seems to be blocked on my templating comment. I'll check it out on my side and then merge if I can't fix it.

perfect- thanks!

**Comment by dellaert** [comment]
I removed @varunagrawal as reviewer. He (should have) no free cycles :-)

**Comment by ProfFan** [comment]
@lucacarlone The CI error seems to be a missing `typedef Measurement` in `Cal3Bundler`

**Comment by lucacarlone** [comment]
> @lucacarlone The CI error seems to be a missing `typedef Measurement` in `Cal3Bundler`

woohoo! this helps a lot!! Feel free to push the fix if you want (I may not be able to get to this before the weekend)

**Comment by dellaert** [comment]
@ProfFan could you help out with CI?

**Comment by lucacarlone** [comment]
@ProfFan : I feel the problem is that there is a missing template in some function call that gets misinterpreted during compilation: If you look at the log, the error refers to using "[with CAMERA = gtsam::Cal3_S2]" which does not make sense since gtsam::Cal3_S2 is a CALIBRATION, not a CAMERA.

**Comment by ProfFan** [comment]
@lucacarlone Let me take a look :)

**Comment by dellaert** [comment]
@lucacarlone , @ProfFan and I diagnosed and fixed the template substitution issue. Now there is only one remaining failing test:
```
 57/264 Test  #58: testTriangulation .............................***Failed    0.07 sec
Not equal:
expected:
[
	0.00180422;
	0.00180422
]
actual:
[
	-0.00180422;
	0.00180421
]
/home/runner/work/gtsam/gtsam/gtsam/geometry/tests/testTriangulation.cpp:605: Failure: "assert_equal(expectedErrorSpherical, actualErrorSpherical, 1e-7)" 
There were 1 failures
```
on some platforms. Maybe two solutions are possible but the choice is not deterministic?

File: gtsam/slam/ReadMe.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
+# SLAM Factors
```
Comment (ProfFan) [inline gtsam/slam/ReadMe.md:1]: This file should be named README.md

File: gtsam_unstable/slam/ReadMe.md [hunk 2]
```diff
@@ -0,0 +1,40 @@
+# SLAM Factors
```
Comment (ProfFan) [inline gtsam_unstable/slam/ReadMe.md:1]: same

File: gtsam/geometry/PinholeCamera.h [hunk 3]
```diff
@@ -312,6 +312,17 @@ class GTSAM_EXPORT PinholeCamera: public PinholeBaseK<Calibration> {
     return range(camera.pose(), Dcamera, Dother);
   }
 
+  /// for Linear Triangulation
+  Matrix34 getCameraProjectionMatrix() const {
```
Comment (dellaert) [inline gtsam/geometry/PinholeCamera.h:316]: Get rid of “get” -> `cameraP…`

File: gtsam/geometry/PinholeCamera.h [hunk 4]
```diff
@@ -312,6 +312,17 @@ class GTSAM_EXPORT PinholeCamera: public PinholeBaseK<Calibration> {
...
+    Matrix34 P = Matrix34(PinholeBase::pose().inverse().matrix().block(0, 0, 3, 4));
```
Comment (dellaert) [inline gtsam/geometry/PinholeCamera.h:317]: Avoid intermediate, just mult with block in return statement (without Matrix34 even)

File: gtsam/geometry/SphericalCamera.h [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+#include <gtsam/dllexport.h>
+#include <gtsam/geometry/BearingRange.h>
+#include <gtsam/geometry/Pose3.h>
+#include <gtsam/geometry/Unit3.h>
+
+#include <boost/serialization/nvp.hpp>
+
+namespace gtsam {
+
+class GTSAM_EXPORT EmptyCal {
```
Comment (dellaert) [inline gtsam/geometry/SphericalCamera.h:33]: Doxygen

File: gtsam/geometry/SphericalCamera.h [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+  EmptyCal() {}
+  virtual ~EmptyCal() = default;
+  using shared_ptr = boost::shared_ptr<EmptyCal>;
+  void print(const std::string& s) const {
+    std::cout << "empty calibration: " << s << std::endl;
+  }
+};
+
+/**
+ * A spherical camera class that has a Pose3 and measures bearing vectors
```
Comment (dellaert) [inline gtsam/geometry/SphericalCamera.h:45]: Note here that it has 
‘Empty’ calibration and the only 6 degrees of freedom are the pose

**Review by dellaert — APPROVED** [review]
Some small nits left, otherwise: amazing!

File: gtsam/geometry/PinholeCamera.h (hunk 3 above)
Comment (lucacarlone) [reply gtsam/geometry/PinholeCamera.h:316]: done

File: gtsam/geometry/PinholeCamera.h (hunk 4 above)
Comment (lucacarlone) [reply gtsam/geometry/PinholeCamera.h:317]: done

File: gtsam/geometry/SphericalCamera.h (hunk 5 above)
Comment (lucacarlone) [reply gtsam/geometry/SphericalCamera.h:33]: done

File: gtsam/geometry/SphericalCamera.h (hunk 6 above)
Comment (lucacarlone) [reply gtsam/geometry/SphericalCamera.h:45]: done

Comment (lucacarlone) [reply gtsam/geometry/SphericalCamera.h:45]: done

**Comment by lucacarlone** [comment]
@dellaert  @ProfFan : this seems to be all set (thanks for your help!!). I'll take a final look and then merge to develop. thanks!

**Pull request merged**